### ISSUE TITLE
How to get a real depth from the depth map?

### ISSUE THREAD

**Issue by wenglihong** [nianticlabs/monodepth2#91]
Hello guys, I'd like to use the depth map to generate point cloud, I ran the test_simple.py and checked the depth output of the network, I found that the depth seemed not in the right scale, for example, the maximum depth in  assets/test_image_disp.jpeg is 4.060234, it's impossible! I added some debug code in test_simple.py and used BackprojectDepth function and the results were the same. So is there some scales in depth estimation and how can I get the right scale of image depth from the program? Here is what I added in test_simple.py and the corresponding output
```
            output_name = os.path.splitext(os.path.basename(image_path))[0]
            name_dest_npy = os.path.join(output_directory, "{}_disp.npy".format(output_name))
            scaled_disp, depth = disp_to_depth(disp, 0.1, 100)
            np.save(name_dest_npy, scaled_disp.cpu().numpy())
            print('max',depth.cpu().numpy().max(), 'min', depth.cpu().numpy().min())
            back_proj = BackprojectDepth(1, 192, 640)
            back_proj.to(device)
            K = np.array([[0.58, 0, 0.5, 0],
                           [0, 1.92, 0.5, 0],
                           [0, 0, 1, 0],
                           [0, 0, 0, 1]], dtype=np.float32)
            K[0, :] *= 640 
            K[1, :] *= 192
            inv_K = np.linalg.pinv(K)
            inv_K = np.expand_dims(inv_K,0)
            inv_K = torch.from_numpy(inv_K).to(device)
            cam_points = back_proj(depth, inv_K)
            print("cam_points", cam_points.shape, cam_points[0, 2, :].max())
```
output information
```
max 4.060234 min 0.17867994
cam_points torch.Size([1, 4, 122880]) tensor(4.0602, device='cuda:0')
```
thanks a lot!

**Comment by aj96** [comment]
@wenglihong The model has no way of learning world scale because it is unsupervised. The scaling values are not consistent accross frames, which is why during evaluation they use the median(gt depth) / median(pred depth) to estimate the scaling value for each example in the test step. If you want the model to learn the depth to metric scale, you will have to use the IMU/GPS data located in the oxts folder to replace the PoseCNN with the ground truth camera-to-camera transformation. I talked about this in this issue: https://github.com/nianticlabs/monodepth2/issues/78

**Comment by mrharicot** [comment]
@wenglihong While the evaluation on KITTI uses a different scale for each test image using the ground truth depth, we evaluated monodepth2 on a single scale for all the images by computing the average of the per sample scales.
You can find this experiment [in the paper](https://arxiv.org/pdf/1806.01260.pdf) in Table 8 of the supplementary material.
This experiment shows that monodepth2's scale estimation is pretty consistent, and even more so than previous methods.
_It is safe to use a single scale for all test samples._

You can find the average scale by running the depth evaluation code:
https://github.com/nianticlabs/monodepth2/blob/master/evaluate_depth.py#L219

You can also use the stereo pretrained model which has its scale known:
https://github.com/nianticlabs/monodepth2/blob/master/evaluate_depth.py#L24

I hope this helps!

**Comment by aj96** [comment]
@mrharicot Just to clarify, each time you evaluate on kitti, you are multiplying the predicted depth by a different scaling factor. You are not using an averaged scale. The line you refered to the average scale is just the median divided by standard deviation of the scale factors. Just printing out the scaling factors, it seems like they vary quite a bit from 37 - 50. They were only truly consistent when including the ground truth localization information.

**Comment by mrharicot** [comment]
@aj96 the method to obtain the scale factor isn't particularly robust, especially due to faraway points, it would be preferable to optimize this value instead.
The line I referred to prints the median  of all the per image scales which you can then use as a parameter with `--disable_median_scaling` and `--pred_depth_scale_factor scale_factor`

**Comment by wenglihong** [comment]
@aj96 @mrharicot thanks for your replies, I read the experiment in the paper and the evaluation code and figure out what's the problem is. I ran the evaluation code and I got these output
`
   Mono evaluation - using median scaling
 Scaling ratios | med: 31.257 | std: 0.089

   abs_rel |   sq_rel |     rmse | rmse_log |       a1 |       a2 |       a3 |
&   0.115  &   0.905  &   4.863  &   0.193  &   0.877  &   0.959  &   0.981  \\
`
I used the median scaling ratio and got the right depth. There is still one question that does the ratio std means the ratios are between 31.257±0.089*3(3 sigma) or it's between 31.257±(31.257*0.089)*3?

**Comment by tareeqav** [comment]
@mrharicot Please forgive my ignorance on this issue as I am still in early days.

When you say "It is safe to use a single scale for all test samples.", does this mean it also generalizes to images from the wild?

Is there any preprocessing required for images taken from an entirely different camera in a different setting? How do the extrinsics of the camera factor into this?

**Comment by mrharicot** [comment]
@wenglihong as you can see in the [code](https://github.com/nianticlabs/monodepth2/blob/master/evaluate_depth.py#L219) the standard deviation is computed on the normalized ratios.

@tareeqav If you want to create a point cloud from a depthmap you need to backproject points into 3d using the camera matrix. I recommend reading up on 3D vision if you don't know how to proceed from here.
Here are some references:
http://web.stanford.edu/class/cs231a/
http://www.computervisionmodels.com/

The network was only trained with KITTI data, it is unclear what it will produce when used on images taken with very different intrinsics. The simple solution would be to start with the scale obtained from the KITTI evaluation and go from there.

**Comment by aj96** [comment]
@tareeqav Yes, you can train it with the GPS/IMU data. Just make sure to disable the random horizontal flipping and to take out noisy examples. For some examples, the altitude would change by 5 meters in just one time-step.

**Comment by ahmeda14960** [comment]
@wenglihong Can you tell me how you ran evaluation code the get the median scaling? I tried looking through the code, and I know that what I want should be something like this:
``` python
for i in range(pred_disps.shape[0]):
        gt_depth = gt_depths[i]
        gt_height, gt_width = gt_depth.shape[:2]

        pred_disp = pred_disps[i]
        pred_disp = cv2.resize(pred_disp, (gt_width, gt_height))
        pred_depth = 1 / pred_disp
        mask = gt_depth > 0

        pred_depth = pred_depth[mask]
        gt_depth = gt_depth[mask]

        pred_depth *= opt.pred_depth_scale_factor
        ratio = np.median(gt_depth) / np.median(pred_depth)
        ratios.append(ratio)
        pred_depth *= ratio
        pred_depth[pred_depth < MIN_DEPTH] = MIN_DEPTH
        pred_depth[pred_depth > MAX_DEPTH] = MAX_DEPTH
```

Which is almost directly from the evaluate_depth file but only focusing on median scaling. However, how exactly are we iterating through a numpy array, because pred_disps is returned by the 'disp_to_depth' function you mentioned above?

**Comment by aj96** [comment]
The median scaling ratio they use is in the code snippet you posted.

`ratio = np.median(gt_depth) / np.median(pred_depth)`

ratio is the median scaling ratio.

**Comment by ahmeda14960** [comment]
@aj96 How do we get the 'gt_depth' variable? If that is the ground truth depth, then can monodepth2 be used on images where we do not have ground truth depth information?

**Comment by aj96** [comment]
@ahmeda14960 I don't think I understand what your ultimate goal is here. This repo provides code for unsupervised monocular-based depth estimation. It learns the pixel-wise depth up to some scaling factor. It does not learn metric depth because monocular-based depth estimation is inherently ambiguous. This scaling factor actually varies quite a bit depending on the image you are looking at. So yes, you can use it where we don't have ground truth depth information, but it will have a varying scaling factor for each image you inference on. If you want to learn to metric depth, you'll have to add the metric information either by adding another loss component that does direct supervision from the ground truth depth or replace the pose cnn with ground truth localization from the IMU/GPS data. These questions have been answered else where on this repo's issues.

**Comment by ahmeda14960** [comment]
@aj96 Thank you for the clarification, I re-read the experiment section in the paper and I now see your point. 

My ultimate goal is to gain some sort of metric depth prediction for an image and to then generate a point cloud similar to @wenglihong. Could you add links to the other repo issues that touch on  using IMU/GPS data to replace the pose cnn?
